contracts/ModuleRegistry.sol

```diff
@@ -3,198 +3,322 @@ pragma solidity ^0.4.24;
             } else {
-                require(verified[_moduleFactory], "ModuleFactory must be verified");
+                require(getBool(Encoder.getKey('verified', _moduleFactory)), "ModuleFactory must be verified");
             }
-            require(registry[_moduleFactory] != 0, "ModuleFactory type should not be 0");
-            reputation[_moduleFactory].push(msg.sender);
-            emit LogModuleUsed(_moduleFactory, msg.sender);
+            uint8[] memory _latestVersion = ISecurityTokenRegistry(getAddress(Encoder.getKey('securityTokenRegistry'))).getProtocolVersion();
+            uint8[] memory _lowerBound = IModuleFactory(_moduleFactory).getLowerSTVersionBounds();
+            uint8[] memory _upperBound = IModuleFactory(_moduleFactory).getUpperSTVersionBounds();
+            require(VersionUtils.compareLowerBound(_lowerBound, _latestVersion), "Version should not be below the lower bound of ST version requirement");
+            require(VersionUtils.compareUpperBound(_upperBound, _latestVersion), "Version should not be above the upper bound of ST version requirement");
+            require(getUint(Encoder.getKey('registry',_moduleFactory)) != 0, "ModuleFactory type should not be 0");
+            pushArray(Encoder.getKey('reputation', _moduleFactory), msg.sender);
+            emit ModuleUsed(_moduleFactory, msg.sender);
         }
     }
 
     /**
-     * @notice Called by moduleFactory owner to register new modules for SecurityToken to use
+     * @notice Called by the ModuleFactory owner to register new modules for SecurityTokens to use
      * @param _moduleFactory is the address of the module factory to be registered
      * @return bool
      */
     function registerModule(address _moduleFactory) external whenNotPaused returns(bool) {
-        require(registry[_moduleFactory] == 0, "Module factory should not be pre-registered");
+        require(getUint(Encoder.getKey('registry', _moduleFactory)) == 0, "Module factory should not be pre-registered");
         IModuleFactory moduleFactory = IModuleFactory(_moduleFactory);
         uint8 moduleType = moduleFactory.getType();
         require(moduleType != 0, "Factory moduleType should not equal to 0");
-        registry[_moduleFactory] = moduleType;
-        moduleListIndex[_moduleFactory] = uint8(moduleList[moduleType].length);
-        moduleList[moduleType].push(_moduleFactory);
-        reputation[_moduleFactory] = new address[](0);
-        emit LogModuleRegistered (_moduleFactory, Ownable(_moduleFactory).owner());
+        set(Encoder.getKey('registry', _moduleFactory), uint256(moduleType));
+        set(Encoder.getKey('moduleListIndex', _moduleFactory), uint256(getArrayAddress(Encoder.getKey('moduleList', uint256(moduleType))).length));
+        pushArray(Encoder.getKey('moduleList', uint256(moduleType)), _moduleFactory);
+        setArray(Encoder.getKey('reputation', _moduleFactory), new address[](0));
```

Comment:
Could you check that we're testing reputation in the test cases? Not sure that we need this line.

Reply:
Yes, we are checking the reputation it is only used for the initialization of address[], not needed it will be performed automatically. So removed from here